src/process/posix/SDL_posixprocess.c

```diff
@@ -0,0 +1,298 @@
+     misrepresented as being the original software.
+  3. This notice may not be removed or altered from any source distribution.
+*/
+#include "SDL_internal.h"
+
+#ifdef SDL_PROCESS_POSIX
+
+#include <fcntl.h>
+#include <errno.h>
+#include <signal.h>
+#include <spawn.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/wait.h>
+
+#include "../SDL_sysprocess.h"
+#include "../../file/SDL_iostream_c.h"
+
+#if defined(SDL_PLATFORM_MACOS)
+#include <crt_externs.h>
+#define environ (*_NSGetEnviron())
+#elif defined(SDL_PLATFORM_FREEBSD)
+#include <dlfcn.h>
+#define environ ((char **)dlsym(RTLD_DEFAULT, "environ"))
+#else
+extern char **environ;
+#endif
+
+#define READ_END 0
+#define WRITE_END 1
+
+struct SDL_ProcessData {
+    pid_t pid;
+};
+
+static void CleanupStream(void *userdata, void *value)
+{
+    SDL_Process *process = (SDL_Process *)value;
+    const char *property = (const char *)userdata;
+
+    SDL_ClearProperty(process->props, property);
+}
+
+static bool SetupStream(SDL_Process *process, int fd, const char *mode, const char *property)
+{
+    // Set the file descriptor to non-blocking mode
+    fcntl(fd, F_SETFL, fcntl(fd, F_GETFL) | O_NONBLOCK);
+
+    FILE *fp = fdopen(fd, mode);
+    if (!fp) {
+        return false;
+    }
+
+    SDL_IOStream *io = SDL_IOFromFP(fp, true);
+    if (!io) {
+        return false;
+    }
+
+    SDL_SetPointerPropertyWithCleanup(SDL_GetIOProperties(io), "SDL.internal.process", process, CleanupStream, (void *)property);
+    SDL_SetPointerProperty(process->props, property, io);
+    return true;
+}
+
+bool SDL_SYS_CreateProcessWithProperties(SDL_Process *process, SDL_PropertiesID props)
+{
+    char * const *args = SDL_GetPointerProperty(props, SDL_PROP_PROCESS_CREATE_ARGS_POINTER, NULL);
+    char * const *env = SDL_GetPointerProperty(props, SDL_PROP_PROCESS_CREATE_ENVIRONMENT_POINTER, environ);
+    bool enable_stdin = SDL_GetBooleanProperty(props, SDL_PROP_PROCESS_CREATE_STDIN_BOOLEAN, false);
+    bool enable_stdout = SDL_GetBooleanProperty(props, SDL_PROP_PROCESS_CREATE_STDOUT_BOOLEAN, false);
+    bool enable_stderr = SDL_GetBooleanProperty(props, SDL_PROP_PROCESS_CREATE_STDERR_BOOLEAN, false);
+    bool redirect_stderr = SDL_GetBooleanProperty(props, SDL_PROP_PROCESS_CREATE_REDIRECT_STDERR_BOOLEAN, false);
+    int stdin_pipe[2] = { -1, -1 };
+    int stdout_pipe[2] = { -1, -1 };
+    int stderr_pipe[2] = { -1, -1 };
+
+    // Keep the malloc() before exec() so that an OOM won't run a process at all
+    SDL_ProcessData *data = SDL_calloc(1, sizeof(*data));
+    if (!data) {
+        return false;
+    }
+    process->internal = data;
+
+    if ((enable_stdin && pipe(stdin_pipe) < 0) ||
+        (enable_stdout && pipe(stdout_pipe) < 0) ||
+        (enable_stderr && pipe(stderr_pipe) < 0)) {
+        SDL_SetError("Could not create pipe(): %s", strerror(errno));
+        goto posix_spawn_fail_none;
+    }
```

Comment:
We should set all of these close-on-execute so that they don't leak into subprocesses. Linux has `pipe2(., O_CLOEXEC)` but this is non-portable, because we can't have nice things. The fallback is to call `pipe()` and then set each of the two fds close-on-execute, something like this (untested, Zlib-licensed):

```
// returns true on success, false with errno set on failure
static bool SDL_SetFdCloseOnExec(int fd)
{
    int flags;

    flags = fcntl(fd, F_GETFD, 0);

    if (flags == -1) {
        return false;
    }

    if (flags & FD_CLOEXEC) {
        return true;
    }

    return fcntl(fd, F_SETFD, flags | FD_CLOEXEC) == 0;
}
```

It would probably be good to have a shared file for Unix-specific utility functions, including `SDL_SetFdCloseOnExec()` and `SDL_OpenPipe()` (which would be equivalent to Linux `pipe2(., O_CLOEXEC)`, but with the fallback code path). Every library that runs subprocesses and is portable ends up with its own copy of this :-(

Reply:
Are the spawn actions not sufficient to close these?

Reply:
The problem is that fds are process-global state which is shared between all threads, so my concern is what happens if another thread of the same process is *also* launching subprocesses. Suppose we have this:

* thread 1 is calling into SDL_Process, and ends up launching process 111
* thread 2 might also be calling into SDL_Process, or might be implementing fork-and-exec itself or via some other library like GLib, Qt or libdbus, and ends up launching process 222

If we don't make these pipes close-on-exec in thread 1, then you're right to say that the spawn actions result in them not being leaked into process 111, but process 222 *will* end up inheriting them, which it wasn't meant to. Meanwhile, if thread 2 doesn't make *its* pipes close-on-exec, then they'll be leaked into process 111.

It's unfortunate that 1980s Unix made inheriting the default and close-on-exec an opt-in, but that ship has sailed (into an iceberg) so we're stuck with it now. Best-practice is to always make every fd except for 0, 1, 2 (stdin, stdout, stderr) be close-on-exec, preferably atomically via O_CLOEXEC, or failing that, as soon as possible after its creation by using fcntl. The only exception is if you are already in a forked subprocess (therefore single-threaded), and you are creating a fd that you specifically *do* want to be carried across the exec that you are about to do.

Reply:
> Meanwhile, if thread 2 doesn't make its pipes close-on-exec, then they'll be leaked into process 111.

On modern glibc we can mitigate that with `posix_spawn_file_actions_addclosefrom_np`, but that's a rather recently-added feature, and is as non-portable as its suffix suggests.

Reply:
👍